controllers/cloudstackfailuredomain_controller.go

```diff
@@ -130,21 +140,222 @@ func (r *CloudStackFailureDomainReconciliationRunner) ClearMachines() (ctrl.Resu
 	if err := r.K8sClient.List(r.RequestCtx, machines, client.MatchingLabels{infrav1.FailureDomainLabelName: r.ReconciliationSubject.Name}); err != nil {
 		return ctrl.Result{}, err
 	}
-	// Deleted CAPI machines for CloudStack machines found.
-	for _, machine := range machines.Items {
+	r.Log.Info(fmt.Sprintf("Clear machine: %d found.", len(machines.Items)))
+	if len(machines.Items) > 0 {
+		if result, err := checkClusterReady(r); err != nil {
+			return result, err
+		}
+	}
+
+	result, err := triggerEtcdClusterRollout(machines.Items, r)
+	if err != nil {
+		return result, err
+	}
+
+	result, err = triggerControlPlaneRollout(machines.Items, r)
+	if err != nil {
+		return result, err
+	}
+
+	result, err = triggerMachineDeploymentRollout(machines.Items, r)
+	if err != nil {
+		return result, err
+	}
+
+	if len(machines.Items) > 0 {
+		return r.RequeueWithMessage("FailureDomain still has machine(s) in it.")
+	}
+	return ctrl.Result{}, nil
+}
+
+func triggerEtcdClusterRollout(machines []infrav1.CloudStackMachine, r *CloudStackFailureDomainReconciliationRunner) (ctrl.Result, error) {
+	var etcdMachine infrav1.CloudStackMachine
+	etcdMachineFound := false
+	for _, machine := range machines {
+		for _, ref := range machine.OwnerReferences {
+			if ref.Kind == "EtcdadmCluster" {
+				etcdMachineFound = true
+				etcdMachine = machine
```

Comment:
why do you need to save the machine? Are you trying to get the etcdOwnerRef? Why not just grab that in here instead of persisting the etcdMachine, then looping through its ownerRefs later on?

Reply:
I combined two loop into one. And also grabbed the ref.